_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,200 @@
+
+{% highlight javascript %}
+import db from './db'
+
+user.validate({
+    id: 'fe6f80c6-c530-4f10-ad0c-89f0f68a879f',
+    email: 'jdoe@email.com',
+    password: 'p@55w0rd'
+})
+// Assumes a db.create method which accepts an object as argument
+// and returns a promise
+.then(db.create).then(res => {
+    console.log('User Created!', res)
+})
+.catch(err => {
+    console.log('Failed!', err)
+})
+{% endhighlight %}
+
+Overall pretty simple stuff. We validate the model, then create the user (or catch the error). But what about that `id`? We probably want to have that generated automatically.
+
+## Creators
+
+Obey supplies `creators` for generating data using methods that can be predefined. For the `id` we'll use the [node-uuid](https://www.npmjs.com/package/node-uuid) module. Let's revisit the model, defining a `creator` using the method by the same name:
+
+{% highlight javascript %}
+import obey from 'obey'
+// Import the node-uuid lib
+import uuid from 'node-uuid'
+
+// Add a creator that returns a v4 UUID if one isn't supplied
+obey.creator('uuidCreator', uuid.v4)
+
+const user = obey.model({
+    // The id now specifies that the creator 'uuid' should be 
+    // used if no value is present
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+So, now we just pass in the `email` and `password` and our `id` is auto-generated.
+
+But that `password` is not cool; we don't want to save that in plain-text...
+
+## Modifiers
+
+Obey wasn't just built for validation, but also to support coercion and modification. Modifiers allow for doing this in a clean, simple method. Let's revisit the model again, using the `modifier` method to create one:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+// Import the argon2 module
+import argon2 from 'argon2'
+
+obey.creator('uuidCreator', uuid.v4)
+
+// Add a modifier that returns md5 encrypted value
+const salt = new Buffer('somesalt')
+obey.modifier('encrypt', val => argon2.hash(val, salt))
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    // Password now specifies the modifier 'encrypt' should be used to
+    // encrypt the password
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Now not only does our validation method create a UUID, it also encrypts the password for us. We've changed the model to accomplish this, but we validate exactly the same, Obey handles this all automatically based on the modelling we defined.
```

Comment:
Last instance of "this" seems to be a _remote reference_. Can it be made more clear?